exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetReaderUtility.java

```diff
@@ -140,15 +140,87 @@ public static int getIntFromLEBytes(byte[] input, int start) {
     return out;
   }
 
+  /**
+   * Map full schema paths in format `a`.`b`.`c` to respective SchemaElement objects.
+   *
+   * @param footer Parquet file metadata
+   * @return       schema full path to SchemaElement map
+   */
   public static Map<String, SchemaElement> getColNameToSchemaElementMapping(ParquetMetadata footer) {
-    HashMap<String, SchemaElement> schemaElements = new HashMap<>();
+    Map<String, SchemaElement> schemaElements = new HashMap<>();
     FileMetaData fileMetaData = new ParquetMetadataConverter().toParquetMetadata(ParquetFileWriter.CURRENT_VERSION, footer);
-    for (SchemaElement se : fileMetaData.getSchema()) {
-      schemaElements.put(se.getName(), se);
+
+    Iterator iter = fileMetaData.getSchema().iterator();
+
+    // skip first default 'root' element
+    if (iter.hasNext()) {
+      iter.next();
+    }
+    while (iter.hasNext()) {
+      addSchemaElementMapping(iter, new StringBuilder(), schemaElements);
     }
     return schemaElements;
   }
 
+  /**
+   * Populate full path to SchemaElement map by recursively traversing schema elements referenced by the given iterator
+   *
+   * @param iter file schema values iterator
+   * @param path parent schema element path
+   * @param schemaElements schema elements map to insert next iterator element into
+   */
+  private static void addSchemaElementMapping(Iterator iter, StringBuilder path,
+      Map<String, SchemaElement> schemaElements) {
+    SchemaElement se = (SchemaElement)iter.next();
+    path.append('`').append(se.getName().toLowerCase()).append('`');
+    schemaElements.put(path.toString(), se);
+
+    int remainingChildren = se.getNum_children();
+
+    while (remainingChildren > 0 && iter.hasNext()) {
```

Comment:
Why we need to count remaining children? Why `iterator.hasNext()` is not sufficient?

Reply:
`iterator.hasNext()` relates to whether there are more elements in flat tree representation and isin itself not enough to maintain correct recursion level. If we don't maintain/check remaining children count for parent element, schema
`a`.`b`
`a`.`c`
will result into keys in the map:
`a`
`a`.`b`
`a`.`b`.`c`
since function will erratically enter recursion where it should not.